shared/gateway/api_middleware.go

```diff
@@ -0,0 +1,463 @@
+			writeError(writer, &DefaultErrorJson{Message: e.Error(), Code: http.StatusInternalServerError}, nil)
+			return
+		}
+
+		var j []byte
+		if data.err.Msg() != "" {
+			// Something went wrong, but the request completed, meaning we can write headers and the error message.
+			for h, vs := range grpcResp.Header {
+				for _, v := range vs {
+					writer.Header().Set(h, v)
+				}
+			}
+			// Set code to HTTP code because unmarshalled body contained gRPC code.
+			data.err.SetCode(grpcResp.StatusCode)
+			writeError(writer, data.err, grpcResp.Header)
+			return
+		} else {
+			// Don't do anything if the response is only a status code.
+			if request.Method == "GET" && data.getResponse != nil {
+				// Deserialize the output of grpc-gateway.
+				if err := json.Unmarshal(body, &data.getResponse); err != nil {
+					e := fmt.Errorf("could not unmarshal response: %w", err)
+					writeError(writer, &DefaultErrorJson{Message: e.Error(), Code: http.StatusInternalServerError}, nil)
+					return
+				}
+				// Apply processing functions to fields with specific tags.
+				if err := processField(data.getResponse, []fieldProcessor{
+					{
+						tag: "hex",
+						f:   base64ToHexProcessor,
+					},
+					{
+						tag: "enum",
+						f:   enumToLowercaseProcessor,
+					},
+					{
+						tag: "time",
+						f:   timeToUnixProcessor,
+					},
+				}); err != nil {
+					e := fmt.Errorf("could not process response hex data: %w", err)
+					writeError(writer, &DefaultErrorJson{Message: e.Error(), Code: http.StatusInternalServerError}, nil)
+					return
+				}
+				// Serialize the return value into JSON.
+				j, err = json.Marshal(data.getResponse)
+				if err != nil {
+					e := fmt.Errorf("could not marshal response: %w", err)
+					writeError(writer, &DefaultErrorJson{Message: e.Error(), Code: http.StatusInternalServerError}, nil)
+					return
+				}
+			}
+		}
+
+		// Write the response (headers + body) and PROFIT!
+		for h, vs := range grpcResp.Header {
+			for _, v := range vs {
+				writer.Header().Set(h, v)
+			}
+		}
+		if request.Method == "GET" {
+			writer.Header().Set("Content-Length", strconv.Itoa(len(j)))
+			writer.WriteHeader(grpcResp.StatusCode)
+			if _, err := io.Copy(writer, ioutil.NopCloser(bytes.NewReader(j))); err != nil {
+				e := fmt.Errorf("could not write response message: %w", err)
+				writeError(writer, &DefaultErrorJson{Message: e.Error(), Code: http.StatusInternalServerError}, nil)
+				return
+			}
+		} else if request.Method == "POST" {
+			writer.WriteHeader(grpcResp.StatusCode)
+		}
+
+		// Final cleanup.
+		if err := grpcResp.Body.Close(); err != nil {
+			e := fmt.Errorf("could not close response body: %w", err)
+			writeError(writer, &DefaultErrorJson{Message: e.Error(), Code: http.StatusInternalServerError}, nil)
+			return
+		}
+	})
+}
+
+// Posted graffiti needs to have length of 32 bytes, but client is allowed to send data of any length.
+func prepareGraffiti(data *endpointData) {
```

Comment:
this package shouldn't know anything about graffiti or actual API items IMO. Maybe there can be a list of callbacks specified on initialization that you can call along the way, and one of them is to handle graffiti and attestations arrays

Reply:
Done exactly that.